src/Catalog/Helpers/CatalogWriterHelper.cs

```diff
@@ -0,0 +1,164 @@
+            IStorage storage,
+            DateTime lastCreated,
+            DateTime lastEdited,
+            DateTime lastDeleted,
+            int maxDegreeOfParallelism,
+            bool? createdPackages,
+            bool updateCreatedFromEdited,
+            CancellationToken cancellationToken,
+            ITelemetryService telemetryService,
+            ILogger logger)
+        {
+            if (packageCatalogItemCreator == null)
+            {
+                throw new ArgumentNullException(nameof(packageCatalogItemCreator));
+            }
+
+            if (packages == null)
+            {
+                throw new ArgumentNullException(nameof(packages));
+            }
+
+            if (storage == null)
+            {
+                throw new ArgumentNullException(nameof(storage));
+            }
+
+            if (maxDegreeOfParallelism < 1)
+            {
+                throw new ArgumentOutOfRangeException(
+                    nameof(maxDegreeOfParallelism),
+                    string.Format(Strings.ArgumentOutOfRange, 1, int.MaxValue));
+            }
+
+            if (telemetryService == null)
+            {
+                throw new ArgumentNullException(nameof(telemetryService));
+            }
+
+            if (logger == null)
+            {
+                throw new ArgumentNullException(nameof(logger));
+            }
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var writer = new AppendOnlyCatalogWriter(storage, telemetryService, Constants.MaxPageSize);
+
+            var lastDate = DetermineLastDate(lastCreated, lastEdited, createdPackages);
+
+            if (packages.Count == 0)
+            {
+                return lastDate;
+            }
+
+            // Flatten the sorted list.
+            var workItems = packages.SelectMany(
+                    pair => pair.Value.Select(
+                        details => new PackageWorkItem(pair.Key, details)))
+                .ToArray();
+
+            await workItems.ForEachAsync(maxDegreeOfParallelism, async workItem =>
+            {
+                workItem.PackageCatalogItem = await packageCatalogItemCreator.CreateAsync(
+                    workItem.FeedPackageDetails,
+                    workItem.Timestamp,
+                    cancellationToken);
+            });
+
+            lastDate = packages.Last().Key;
+
+            // AppendOnlyCatalogWriter.Add(...) is not thread-safe, so add them all at once on one thread.
+            foreach (var workItem in workItems.Where(workItem => workItem.PackageCatalogItem != null))
+            {
+                writer.Add(workItem.PackageCatalogItem);
+
+                logger?.LogInformation("Add metadata from: {PackageDetailsContentUri}", workItem.FeedPackageDetails.ContentUri);
+            }
+
+            if (createdPackages.HasValue)
+            {
+                lastEdited = !createdPackages.Value ? lastDate : lastEdited;
```

Comment:
Should be createdPackages.Value instead of !createdPackages.Value?  There are not any checks for the order of lastEdited > or < lastDate . Is any logic needed for this?

Reply:
This is existing code which just moved to another type.